ConfirmPage: update UI for setApproveForAll transactions to better warn users of the allowances

digiwand opened the pull request:
## Explanation

In https://github.com/MetaMask/metamask-extension/pull/15496, @cygaar proposed improving the warning message on the ConfirmPage for `setApprovalForAll` transaction types and gathered feedback to support this improvement.

Based on an internal dialog [here](https://consensys.slack.com/archives/GTQAGKY5V/p1659966850991299?thread_ts=1659958204.365809&cid=GTQAGKY5V), we will be updating the "Confirm" button to red, and including an error type Dialog to better warn users that they are interacting with a `setApprovalForAll` transaction. This is a temporary improvement while a redesign is on the way. More details are explained in https://github.com/MetaMask/metamask-extension/pull/15496 and the [internal dialog](https://consensys.slack.com/archives/GTQAGKY5V/p1659966850991299?thread_ts=1659958204.365809&cid=GTQAGKY5V).

## More Information

See: https://github.com/MetaMask/metamask-extension/pull/15496
See: [Internal Dialog](https://consensys.slack.com/archives/GTQAGKY5V/p1659966850991299?thread_ts=1659958204.365809&cid=GTQAGKY5V)

## Screenshots/Screencaps
### Before
<img width="468" alt="Screenshot 2022-08-10 at 3 17 20 AM" src="https://user-images.githubusercontent.com/20778143/183743004-23ee911a-8393-4e83-a3c7-0822171905dc.png">

### After
<img width="526" alt="Screenshot 2022-08-10 at 6 06 42 PM" src="https://user-images.githubusercontent.com/20778143/183882919-ca3f15b4-dd5e-4751-a9bb-07d91567fa30.png">
<img width="406" alt="Screenshot 2022-08-10 at 6 07 04 PM" src="https://user-images.githubusercontent.com/20778143/183882944-4ee92462-b118-48f5-91f1-604bc60876a0.png">


## Manual Testing Steps

- send a `setApprovalForAll` transaction and observe a red, danger-primary "Confirm" button
- send any other transaction and observe a blue, primary "Confirm" button

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

digiwand committed:
ConfirmPage: setApproveForAll use danger-primary

digiwand committed:
ConfirmPageContainer: add setApprovalForAll dialog
- https://consensys.slack.com/archives/GTQAGKY5V/p1660061693528129
- temp UI to add protections for user until new redesign is implemented

cygaar commented:
Great to see, thanks for updating!

digiwand committed:
Merge branch 'develop' into style/setApprovalForAll-red-confirm-btn

georgewrmarshall commented on app/_locales/en/messages.json:
```diff
@@ -596,6 +596,10 @@
     "message": "Confirm"
   },
+  "confirmPageDialogSetApprovalForAll": {
+    "message": "You’re allowing access to all of your $1 NFTs plus any future NFTs in this collection. This should never be your default, no matter how secure a site may seem. Proceed with caution.",
```
I believe our general content guidelines are to use punctuation for any sentences that have proceeding sentences but not at the end of the last sentence? @coreyjanssen can you confirm?

```suggestion
    "message": "You’re allowing access to all of your $1 NFTs plus any future NFTs in this collection. This should never be your default, no matter how secure a site may seem. Proceed with caution",
```

adonesky1 commented:
The language: `This should never be your default` might throw some users off, since almost no dapps will present users with a secondary option (say, to approve a single tokenId). I would find this language confusing. I think we should make the risks known and present it with the appropriate gravity of warning without suggesting that the user is missing some less risky option that very mainstream dapps likely don't offer

adonesky1 reviewed (changes requested):
The language `This should never be your default` included in the warning might throw some users off, since almost no dapps will present users with a secondary option (say, to approve a single tokenId). I would find this language confusing. I think we should make the risks known and present it with the appropriate gravity of warning without suggesting that the user is missing some less risky option that very mainstream dapps likely don't offer. Open to being convinced otherwise though! cc @tayvano tagging you since I know you have strong opinions about security and the language we use to convey it.

georgewrmarshall commented on ui/components/app/confirm-page-container/index.scss:
```diff
@@ -3,6 +3,12 @@
 
+.confirm-page-container {
+  &__dialog {
+    margin: 0.75rem 1rem;
```
I've been pushing for using pixel units for bounding box values just to save some cognitive load from Figma to code. There is a great article by Josh Comeau on this https://www.joshwcomeau.com/css/surprising-truth-about-pixels-and-accessibility/

```suggestion
    margin: 12px 16px;
```

georgewrmarshall reviewed (commented):
Looks great! A couple tiny nit suggestions

bschorchit commented:
> The language This should never be your default included in the warning might throw some users off, since almost no dapps will present users with a secondary option (say, to approve a single tokenId). I would find this language confusing. I think we should make the risks known and present it with the appropriate gravity of warning without suggesting that the user is missing some less risky option that very mainstream dapps likely don't offer. Open to being convinced otherwise though! cc @tayvano tagging you since I know you have strong opinions about security and the language we use to convey it.

@adonesky1 what do you think about: "You're giving access to all your current and future <NFTname> NFTs. The other party might be able to transfer them from your wallet without further notice. Proceed with caution."
cc: @SayaGT @coreyjanssen

adonesky1 commented:
How about: "Bu clicking approve you are granting access to all of the NFTs you currently own on this contract and any NFTs on this contract you may acquire in the future, until you revoke this approval. The party to whom you're granting approval will be able to transfer your NFTs from your wallet without further notice. Proceed with caution" something like that... away from my computer right not so grammar could use some tweaking

bschorchit commented:
> How about: "By clicking approve you are granting access to all of the NFTs you currently own on this contract and any NFTs on this contract you may acquire in the future, until you revoke this approval. The party to whom you're granting approval will be able to transfer your NFTs from your wallet without further notice. Proceed with caution" something like that... away from my computer right not so grammar could use some tweaking

Sounds great to me!

digiwand committed:
ConfirmPageContainer: use px instead of rem

Co-authored-by: George Marshall <george.marshall@consensys.net>

digiwand committed:
ConfirmPageContainer: update SetApprovalForAll msg

digiwand committed:
Merge branch 'style/setApprovalForAll-red-confirm-btn' of github.com:MetaMask/metamask-extension into style/setApprovalForAll-red-confirm-btn

digiwand commented:
Great call out and suggestion @adonesky1! Looks good to me. The first message was copied from a mock supplied in this [Slack thread](https://consensys.slack.com/archives/GTQAGKY5V/p1659966850991299?thread_ts=1659958204.365809&cid=GTQAGKY5V). 

I moved one comma and kept the rest the same:
> By clicking approve, you are granting access to all of the NFTs you currently own on this contract and any NFTs on this contract you may acquire in the future until you revoke this approval. The party to whom you're granting approval will be able to transfer your NFTs from your wallet without further notice. Proceed with caution.

Updated: https://github.com/MetaMask/metamask-extension/pull/15512/commits/5e0b148eee0a4d1a2a34366c7ebac34e7172fd8e

-- 

In regards to keeping or omitting the punctuation on the last sentence, we could do that in a follow-up PR if this PR gets approved before the final decision here. ref: https://github.com/MetaMask/metamask-extension/pull/15512#discussion_r941765206

digiwand committed:
Merge branch 'develop' into style/setApprovalForAll-red-confirm-btn

adonesky1 reviewed (approved):
LGTM! Thanks @digiwand for picking up the contributor feedback here and moving this forward!

coreyjanssen commented:
Apologies for not seeing this convo until today! (thanks @bschorchit!) Made a few edits to the proposed copy. Lmk if this works!

You're granting access to **all the NFTs on this contract**, including any you might own in the future. The party on the other end of this approval can transfer NFTs from your wallet at any time without asking you. **Proceed with caution.**

digiwand commented:
Hi @coreyjanssen, 

No problem! I like this update, and I'm up for updating this. One thing that I would miss from the copy that has been merged
> By clicking approve, you are granting access to all of the NFTs you currently own on this contract and any NFTs on this contract you may acquire in the future until you revoke this approval. The party to whom you're granting approval will be able to transfer your NFTs from your wallet without further notice. Proceed with caution.

is the mention that this contract is able to access the NFTs `until you revoke this approval.`. Is this something we could add in somehow? 

cc: @adonesky1 @bschorchit

coreyjanssen commented:
@digiwand sure thing! how's this?

You're granting access to **all the NFTs on this contract**, including any you might own in the future. The party on the other end can transfer NFTs from your wallet at any time without asking you until you revoke this approval. **Proceed with caution.**

digiwand commented:
Awesome!!! I think this is an A1 update! Thanks @coreyjanssen! :D

Updated here: https://github.com/MetaMask/metamask-extension/pull/15744
